### PR TITLE
Introduce Espresso tests framework and several general tests

### PR DESCRIPTION
**Pull request by werman**
It's time to add some tests.
Our code was not made with unit tests in mind and most of errors are in
interaction between components so testing UI looks like a way to go.

Each test will be launched with clean state due to the usage of Orchestrator.

### CODE REVIEW

**Comment by werman** [comment]
There are some rough edges at the moment:

1) Recycler view with stations does not really end at the start of mini player, mini player does cover bottom, of the recycler, this messes up with drag-and-drop tests.
2) Network requests do not return predetermined results and are real. I included necessary lib but didn't use it.

**Comment by segler-alex** [comment]
cool work, thank you for adding tests!!!

**Comment by werman** [comment]
@morckx I made your unit tests use JUnit4 since espresso tests don't support JUnit5 and I had troubles using them together.

There are 2-3 unstable tests at the moment, I'll think what to do with them. Otherwise it's almost done.

Tests aren't fast, there is a long setup on CI, and then each test relaunches application, so we should be careful adding tests.

**Comment by werman** [comment]
@morckx I think the tests are ready, could you take a look at them?

Current issues:
- Rare (1/6) failure of `stationInHistory_ShouldBeDeleted_WithSwipeRight`, snackbar sometimes isn't shown for some unknown for me reason.
- Rare failures in `UINotificationTests` subtests - not sure why.
- Drag-and-drop tests are disabled due to https://stackoverflow.com/questions/27992427/recyclerview-adapter-notifyitemmoved0-1-scrolls-screen

TODO (not in this PR):
- We could batch some tests together - at the moment teach test clears and relaunches application, some tests are safe to be batched, I think... See https://chromium.googlesource.com/chromium/src/+/master/testing/android/docs/instrumentation.md#test-batching (it's not implemented out of the box in espresso).
- Test with [app crawler](https://developer.android.com/training/testing/crawler) - not strictly related to espresso, but I have to mention it.
- And of course more UI tests.

What I have learnt:
- As is with many google's products - espresso is half-baked, out-of-touch with real developers' needs:
  - Testing drag-and-drop - write your own helpers
  - Testing recyclers - write your own helpers
  - You want to open notification - ok. You want to close them? NO
  - You want to launch this specific test from AndroidStudio - how about NO
  - You want to understand why there is no tests found - worm through many layers of the framework, finding that it thinks that the test is being run on a desktop...
  - And so on

**Comment by morckx** [comment]
> @morckx I think the tests are ready, could you take a look at them?

Great! I'll have a look.

File: app/src/androidTest/java/net/programmierecke/radiodroid2/tests/UIFavouritesFragmentTest.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    @Test
+    public void stationsRecyclerFavourites_ShouldRecycleItems() {
+        onView(ViewMatchers.withId(R.id.nav_item_starred)).perform(ViewActions.click());
+
+        onView(withId(R.id.recyclerViewStations)).check(matches(recyclerRecycles()));
+    }
+
+    @Ignore("Disabled until drag and drop is fixed, see " +
+            "https://stackoverflow.com/questions/27992427/recyclerview-adapter-notifyitemmoved0-1-scrolls-screen")
```
[the remark this answers is not in the document]
Comment (werman) [reply app/src/androidTest/java/net/programmierecke/radiodroid2/tests/UIFavouritesFragmentTest.java:86]: Agree with drag-and-drop.

> androidx.test.espresso.AmbiguousViewMatcherException

I never had this error on these tests. Something funny is going on here. And probably would need a proper solution.

File: app/build.gradle [hunk 2]
```diff
@@ -1,5 +1,4 @@
 apply plugin: 'com.android.application'
-apply plugin: "de.mannodermaus.android-junit5"
```
[the remark this answers is not in the document]
Comment (werman) [reply app/build.gradle:2]: I didn't have issues running tests without it.

File: .github/workflows/android-instrumented-tests-ci.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  pull_request:
+    branches:
+      - master
+
+jobs:
+  integration-test:
+    runs-on: macOS-10.14
+    strategy:
+      matrix:
+        api: [16, 21, 27, 29]
```
[the remark this answers is not in the document]
Comment (werman) [reply .github/workflows/android-instrumented-tests-ci.yml:16]: Hmmm, I never saw this issues...

**Comment by morckx** [comment]
> 
> What I have learnt:
> 
> * As is with many google's products - espresso is half-baked, out-of-touch with real developers' needs:

I can confirm that in this case. I had already contributed some espresso tests to k-9 mail and connectbot. There my experiences were much more positive.

Anyway, I think this is definitely a good start we can build upon. Thanks!
 
>   * You want to launch this specific test from AndroidStudio - how about NO

That's really annoying.

>   * You want to understand why there is no tests found - worm through many layers of the framework, finding that it thinks that the test is being run on a desktop...

:)

File: .github/workflows/android-instrumented-tests-ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (morckx) [reply .github/workflows/android-instrumented-tests-ci.yml:16]: Me also only with the github workflow and only occasionally – I think also with one more language, but I'm no sure.

File: .github/workflows/android-instrumented-tests-ci.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      - uses: malinskiy/action-android/install-sdk@release/0.0.6
+      - name: Instrumented tests
+        uses: malinskiy/action-android/emulator-run-cmd@release/0.0.6
+        timeout-minutes: 35
+        with:
+          cmd: ./gradlew connectedFreeDebugAndroidTest
+          api: ${{ matrix.api }}
+          tag: default
+          abi: x86
+          hardwareProfile: Nexus 6
```
Comment (morckx) [inline .github/workflows/android-instrumented-tests-ci.yml:32]: `disableAnimations: true` seems to be essential here, see https://github.com/morckx/RadioDroid/commit/27cacd884b892ee02f52dac89c2ec6a836f6cd1c

File: app/src/androidTest/java/net/programmierecke/radiodroid2/tests/UIFavouritesFragmentTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (morckx) [reply app/src/androidTest/java/net/programmierecke/radiodroid2/tests/UIFavouritesFragmentTest.java:86]: I didn't get this error anymore, either.

File: app/src/androidTest/java/net/programmierecke/radiodroid2/tests/UINotificationTests.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+import static androidx.test.espresso.Espresso.onView;
+import static androidx.test.espresso.matcher.ViewMatchers.isDisplayingAtLeast;
+import static androidx.test.espresso.matcher.ViewMatchers.withId;
+import static org.hamcrest.core.AllOf.allOf;
+import static org.junit.Assert.assertNotNull;
+
+@LargeTest
+@RunWith(AndroidJUnit4.class)
+// UiDevice can be used only with API 18+
+@SdkSuppress(minSdkVersion = 18)
```
[the remark this answers is not in the document]
Comment (werman) [reply app/src/androidTest/java/net/programmierecke/radiodroid2/tests/UINotificationTests.java]: They have description, on my emulators they fail in 25% cases, but I don't know even how to debug. I feel that debugging graphics drivers sometimes it easier than android...

**Comment by werman** [comment]
@morckx Thank you for making the final changes!

**Pull request merged**